cli/docker/container_manager.go

```diff
@@ -0,0 +1,545 @@
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Error(redacted)
+}
+
+func (dl *DockerLogger) Warnf(format string, args ...interface{}) {
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Warn(redacted)
+}
+
+func (dl *DockerLogger) Redact(s string) (string, error) {
+	src := bytes.NewReader([]byte(s))
+	dstBuf := new(bytes.Buffer)
+	logger := log.NewUtilsLogAdapter()
+	redactWriterDst := redactwriter.New(dl.secrets, dstBuf, &logger)
+
+	if _, err := io.Copy(redactWriterDst, src); err != nil {
+		return "", fmt.Errorf("failed to redact secrets, stream copy failed: %s", err)
+	}
+	if err := redactWriterDst.Close(); err != nil {
+		return "", fmt.Errorf("failed to redact secrets, closing the stream failed: %s", err)
+	}
+
+	redactedValue := dstBuf.String()
+	return redactedValue, nil
+}
+
+func NewContainerManager(logger log.Logger, secrets []string) *ContainerManager {
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		logger.Warnf("Docker client failed to initialize (possibly running on unsupported stack): %s", err)
+	}
+
+	return &ContainerManager{
+		logger: DockerLogger{
+			logger:  logger,
+			secrets: secrets,
+		},
+		workflowContainers: make(map[string]*RunningContainer),
+		serviceContainers:  make(map[string][]*RunningContainer),
+		client:             dockerClient,
+	}
+}
+
+func (cm *ContainerManager) Login(container models.Container, envs map[string]string) error {
+	if container.Credentials.Username != "" && container.Credentials.Password != "" {
+		cm.logger.Infof("ℹ️ Logging into docker registry: %s", container.Image)
+		cm.logger.Infof("ℹ️ Logging into docker registry: %s", container.Image)
+
+		resolvedPassword := resolveEnvVariable(container.Credentials.Password, envs)
+		args := []string{"login", "--username", container.Credentials.Username, "--password", resolvedPassword}
+
+		if container.Credentials.Server != "" {
+			args = append(args, container.Credentials.Server)
+		} else {
+			args = append(args, container.Image)
+		}
+
+		cm.logger.Infof("ℹ️ Running command: docker %s", strings.Join(args, " "))
+
+		out, err := command.New("docker", args...).RunAndReturnTrimmedCombinedOutput()
+		if err != nil {
+			cm.logger.Errorf(out)
+			return fmt.Errorf("run docker login: %w", err)
+		}
+	}
+	return nil
+}
+
+func (cm *ContainerManager) StartWorkflowContainer(
+	container models.Container,
+	workflowID string,
+	envs map[string]string,
+) (*RunningContainer, error) {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	containerName := fmt.Sprintf("workflow-%s", workflowID)
+	dockerMountOverrides := strings.Split(os.Getenv("BITRISE_DOCKER_MOUNT_OVERRIDES"), ",")
+
+	runningContainer, err := cm.runContainer(container, containerCreateOptions{
+		name:       containerName,
+		volumes:    dockerMountOverrides,
+		command:    "sleep infinity",
+		workingDir: "/bitrise/src",
```

Comment:
Is this a good idea? It exists on the current Linux stack, but there is no guarantee that it won't change in the future. Maybe resolving the current working dir or `$BITRISE_WORKING_DIR` would be better.

Reply:
If it changes significantly it could break the whole functionality as this folder must be properly mounted. We could change this to a variable but it must be synched with DEN either ways anytime this changes.

Reply:
Note to ourselves (cc: @bvatai-br): We called this dockerMountOverrides because we meant this to override some defaults, but we have not thought about that just yet. So this particular version will not work out of the box when someone were to run bitrise-cli on their local machine using container workflows.

It is not trivial to do a default here as it depends on the operating system (mac vs linux), docker engine settings/permissions, but maybe we could try to create one and depending on the detected OS default to that one.

This could be an improvement to this PR later as local builds are not the primary focus at this time, but nevertheless  we need to keep this in mind.